n3fit/src/n3fit/model_trainer.py

```diff
@@ -289,7 +292,60 @@ def _fill_the_dictionaries(self):
                 self.training["expdata"].append(integ_dict["expdata"])
                 self.training["integdatasets"].append(integ_dict["name"])
 
-    def _model_generation(self, pdf_models, partition, partition_idx):
+    def _xgrid_generation(self):
+        """
+        Generates the full x-grid pertaining to the complete set of observables to be fitted.
+
+        To first approximation, the full x-grid is a concatenation of all x-grid requested by
+        all fk-tables.
+
+        In the case of pineappl models all fktables ask for the same grid in x
+        and so the input can be simplified to be a single grid for all (or most) datasets.
+        However, this is not a _strict_ requirement for pineappl and was not a requirement before
+        so the solution below must be kept general enough.
+
+        Detailed implementation of the union of xgrids:
+            let's assume an input [x1, x1, x1, x2, x2, x3]
+            where each xi is a different grid, this will be broken into two lists:
+            [x1, x2, x3] (unique grids) and [0,0,0,1,1,2] (index of the grid per dataset)
+            The pdf will then be evaluated to concatenate([x1,x2,x3]) and then split (x1, x2, x3)
+            Then each of the experiment, looking at the indexes, will receive one of the 3 PDFs
+            The decision whether two grids (x1 and x1) are really the same is decided below
+
+        The necessary information to redistribute the x-grid is held by a ``InputInfo`` tuple
```

Comment:
Perhaps we could have a Returns field here explaining each element of the tuple?